### ISSUE TITLE
Add patch as suggestion to PR

### ISSUE THREAD

**Issue by slarse** [SpoonLabs/sorald-buildbreaker#53]
When the build breaks, add patch as suggestion

**Comment by slarse** [comment]
It looks totally possible to do with review comments: https://docs.github.com/en/rest/reference/pulls#review-comments

FUN!

**Comment by slarse** [comment]
This seems very straightforward. This little code snippet posts a review comment with a suggestion.

```typescript
  const octokit = new Octokit({ auth: `SUPER_SECRET_TOKEN` });
  const response = await octokit.rest.pulls.createReviewComment({
    owner: 'spoonlabs',
    repo: 'sorald',
    commit_id: 'f88c94e4697bbfb40f782a3c6365861db3db0d71',
    pull_number: 534,
    body: '```suggestion\nthis is better\n```',
    path: 'src/main/java/sorald/Main.java',
    start_line: 7,
    line: 12,
    start_side: 'RIGHT'
  });
  console.log(response);
```

Here's the comment that was actually posted: https://github.com/SpoonLabs/sorald/pull/534#discussion_r633542017

**Comment by slarse** [comment]
So, I'm satisfied in that this is totally doable in terms of the GitHub API features. Before we can make this happen, we need a way to extract some more information from diff hunks, opening a new issue for that.

**Comment by algomaster99** [comment]
Sounds like a good idea! The implementation is also nice. But I am assuming that the POST request would require some kind of token in case the repository with which buildbreaker is integrated is private?

**Comment by slarse** [comment]
A GitHub API token is always available to workflows. When a PR comes from a fork, that token is however read-only, which may pose problems for external pull requests, but I _think_ it should still allow to post a comment. Haven't tried it yet though. But this is a prototype at this stage, so it doesn't really matter too much, we'll figure it out as we go along.

As for private repositories, the same things apply there. And we aren't really concerned with private repos at this point, if there's anything in particular we need to add for private repos and someone actually requests it, we'll consider it then.

**Comment by algomaster99** [comment]
> external pull requests

What do you mean by "external pull requests"?

> if there's anything in particular we need to add for private repos and someone actually requests it, we'll consider it then.

Yeah, sure. I was just curious about the GitHub APIs functioning. I worked with them 2 years ago.

**Comment by slarse** [comment]
> > external pull requests
> 
> What do you mean by "external pull requests"?

PR from a fork, i.e. not from within the same repository. The API token is then limited to read-only, and secrets are unavailable.

**Comment by algomaster99** [comment]
Oh, okay. Yeah, I think it should allow to post a comment. The APIs should replicate the GitHub's interface behaviour in my opinion.

**Comment by slarse** [comment]
> The APIs should replicate the GitHub's interface behaviour in my opinion.

They do, and more to boot. But it's a matter of permissions: can a read-only token _write_ a comment?

After briefly trying it out, it appears that it cannot. Which is reasonable, considering it being read-only. It's possible sorald-buildbreaker will have to be turned into a GitHub bot to fully realize it, but we can still continue with this prototype.

**Comment by algomaster99** [comment]
> It's possible sorald-buildbreaker will have to be turned into a GitHub bot to fully realize it

Oh, okay. I was quite looking forward to the patches when I break build in `diffmin`. 😅

**Comment by slarse** [comment]
> Oh, okay. I was quite looking forward to the patches when I break build in diffmin. sweat_smile

That will still work, because your PRs are internal to the repository. They don't come from forks.

**Comment by algomaster99** [comment]
Oh, right. I lost track that we were talking about the external pull requests _only_.
